onnxruntime/core/providers/dml/DmlExecutionProvider/src/Operators/DmlOperatorAttention.cpp

```diff
@@ -0,0 +1,401 @@
+        MLOperatorTensorDataType dataType = kernelCreationContext.GetInputEdgeDescription(0).tensorDataType;
+
+        // overwrite weightTensorDesc
+        m_inputTensorDescs[1] = TensorDesc::ConstructDefaultTensorDesc(dataType, desiredWeightTensorShape, weightTensorShape);
+
+        // overwrite biasTensorDesc
+        m_inputTensorDescs[2] = TensorDesc::ConstructDefaultTensorDesc(dataType, desiredBiasTensorShape, biasTensorShape);
+
+        // overwrite maskIndexTensorDesc
+        uint32_t maskIndexDimensionCount = gsl::narrow_cast<uint32_t>(maskIndexTensorShape.size());
+        maskIndexTensorShape.insert(maskIndexTensorShape.begin() + 1, 4 - maskIndexDimensionCount, 1);
+        uint32_t desiredMaskIndexShape[4] {batchSize, numHeads, sequenceLength, sequenceLength};
+        MLOperatorTensorDataType maskTensorDataType = kernelCreationContext.GetInputEdgeDescription(3).tensorDataType;
+        m_inputTensorDescs[3] = TensorDesc::ConstructDefaultTensorDesc(maskTensorDataType, desiredMaskIndexShape, maskIndexTensorShape);
+
+        // overwrite output tensor desc
+        uint32_t outputTensorShape[4] {batchSize, sequenceLength, numHeads, headSize};
+        uint32_t outputTensorStrides[4] {sequenceLength * numHeads * headSize, headSize, headSize * sequenceLength, 1};
+        m_outputTensorDescs[0] = TensorDesc(GetDmlDataTypeFromMlDataType(dataType), outputTensorShape, outputTensorStrides, 0);
+
+        TensorDesc firstGemmOutputTensorDesc = TensorDesc::ConstructDefaultTensorDesc(dataType, desiredBiasTensorShape, desiredBiasTensorShape);
+        DML_TENSOR_DESC namedFirstGemmOutputTensorDesc = firstGemmOutputTensorDesc.GetDmlDesc();
+
+        std::vector<DML_TENSOR_DESC> inputDescs = GetDmlInputDescs();
+        std::vector<DML_TENSOR_DESC> outputDescs = GetDmlOutputDescs();
+
+        DML_GEMM_OPERATOR_DESC xWeightOperatorDesc = {};
+        xWeightOperatorDesc.ATensor = &inputDescs[0];
+        xWeightOperatorDesc.BTensor = &inputDescs[1];
+        xWeightOperatorDesc.CTensor = &inputDescs[2];
+        xWeightOperatorDesc.OutputTensor = &namedFirstGemmOutputTensorDesc;
+        xWeightOperatorDesc.TransA = DML_MATRIX_TRANSFORM_NONE;
+        xWeightOperatorDesc.TransB = DML_MATRIX_TRANSFORM_NONE;
+        xWeightOperatorDesc.Alpha = 1.0f;
+        xWeightOperatorDesc.Beta = 1.0f;
+        xWeightOperatorDesc.FusedActivation = nullptr;
+        const DML_OPERATOR_DESC xWeightDesc {DML_OPERATOR_GEMM, &xWeightOperatorDesc};
+
+
+        std::array<uint32_t, 3> querySlicedTensorShape {batchSize, sequenceLength, hiddenSize};
+        TensorDesc querySlicedInputTensorDesc = TensorDesc::ConstructDefaultTensorDesc(dataType, querySlicedTensorShape, querySlicedTensorShape);
+        DML_TENSOR_DESC namedQuerySlicedInputTensorDesc = querySlicedInputTensorDesc.GetDmlDesc();
+
+        std::array<uint32_t, 3> querySliceOffset {0, 0, 0}, keySliceOffset {0, 0, hiddenSize}, valueSliceOffset {0, 0, 2 * hiddenSize};
```

Comment:
>keySliceOffset {0, 0, hiddenSize}, valueSliceOffset {0, 0, 2 * hiddenSize}; [](http://example.com/codeflow?start=60&length=75)

request, would you put these on separate lines?
It usually makes it easier to see where blocks of variables are defined when jumping around the code.

```
std::array<uint32_t, 3> querySliceOffset = {0, 0, 0};
std::array<uint32_t, 3> keySliceOffset = {0, 0, hiddenSize};
std::array<uint32_t, 3> valueSliceOffset = {0, 0, 2 * hiddenSize};
```

Reply:
request acknowledged. :smile: